tests/integration/test_clip.py

```diff
@@ -0,0 +1,147 @@
+        f.post(on='/test', inputs=[Document(blob=np.ones((3, 100, 100)))], on_done=validate_callback)
+
+
+def test_clip_batch():
+    def validate_callback(resp):
+        assert 25 == len(DocumentArray(resp.data.docs).get_attributes('embedding'))
+
+    f = Flow().add(uses={
+        'jtype': CLIPImageEncoder.__name__,
+        'with': {
+            'default_batch_size': 10,
+            'model_name': 'ViT-B/32',
+            'device': 'cpu'
+        }
+    })
+    with f:
+        f.post(on='/test', inputs=(Document(blob=np.ones((3, 224, 224))) for _ in range(25)), on_done=validate_callback)
+
+
+def test_traversal_path():
+    def validate_default_traversal(resp):
+        for path, count in [['r', 0], ['c', 3], ['cc', 0]]:
+            assert len(DocumentArray(resp.data.docs).traverse_flat([path]).get_attributes('embedding')) == count
+
+    def validate_request_traversal(resp):
+        for path, count in [['r', 0], ['c', 0], ['cc', 2]]:
+            assert len(DocumentArray(resp.data.docs).traverse_flat([path]).get_attributes('embedding')) == count
+
+    blob = np.ones((3, 224, 224))
+    docs = [Document(
+        id='root1',
+        blob=blob,
+        chunks=[
+            Document(
+                id='chunk11',
+                blob=blob,
+                chunks=[
+                    Document(id='chunk111', blob=blob),
+                    Document(id='chunk112', blob=blob),
+                ]
+            ),
+            Document(id='chunk12', blob=blob),
+            Document(id='chunk13', blob=blob),
+        ]
+    )]
+    f = Flow().add(uses={
+        'jtype': CLIPImageEncoder.__name__,
+        'with': {
+            'default_traversal_path': ['c'],
+            'model_name': 'ViT-B/32',
+            'device': 'cpu'
+        }
+    })
+    with f:
+        f.post(on='/test', inputs=docs, on_done=validate_default_traversal)
+        f.post(on='/test', inputs=docs, parameters={'traversal_path': ['cc']}, on_done=validate_request_traversal)
+
+
+def test_custom_processing():
+    f = Flow().add(uses=CLIPImageEncoder)
+    with f:
+        result1 = f.post(on='/test', inputs=[Document(blob=np.ones((3, 224, 224)))], return_results=True)
+
+    f = Flow().add(uses={
+        'jtype': CLIPImageEncoder.__name__,
+        'with': {
+            'use_default_preprocessing': False,
+        }
+    })
+
+    with f:
+        result2 = f.post(on='/test', inputs=[Document(blob=np.ones((3, 224, 224)))], return_results=True)
+
+    assert result1[0].docs[0].embedding is not None
+    assert result2[0].docs[0].embedding is not None
+    np.testing.assert_array_compare(operator.__ne__, result1[0].docs[0].embedding, result2[0].docs[0].embedding)
+
+
+def test_clip_data():
+    docs = []
+    for file in glob(os.path.join(cur_dir, 'data', '*')):
+        pil_image = Image.open(file)
+        nd_image = np.array(pil_image)
+        prepared_nd_image = np.moveaxis((nd_image / 255), -1, 0)
```

Comment:
I think this is a symptom that the preprocessing is assuming something wrong?

Reply:
yes, you are right. The user might expect just the (h, w, 3) format.

Reply:
changed